kernel/src/transaction.rs

```diff
@@ -0,0 +1,543 @@
+    Conflict(Transaction, Version),
+}
+
+// given the engine's commit info we want to create commitInfo action to commit (and append more actions to)
+fn generate_commit_info(
+    engine: &dyn Engine,
+    operation: Option<&str>,
+    engine_commit_info: &dyn EngineData,
+) -> DeltaResult<Box<dyn EngineData>> {
+    if engine_commit_info.length() != 1 {
+        return Err(Error::InvalidCommitInfo(format!(
+            "Engine commit info should have exactly one row, found {}",
+            engine_commit_info.length()
+        )));
+    }
+
+    let timestamp: i64 = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("time went backwards..?")
+        .as_millis()
+        .try_into()
+        .map_err(|_| Error::generic("milliseconds since unix_epoch exceeded i64 size"))?;
+    let commit_info_exprs = [
+        // TODO(zach): we should probably take a timestamp closer to actual commit time?
+        Expression::literal(timestamp),
+        Expression::literal(operation.unwrap_or(UNKNOWN_OPERATION)),
+        // HACK (part 1/2): since we don't have proper map support, we create a literal struct with
+        // one null field to create data that serializes as "operationParameters": {}
+        Expression::literal(Scalar::Struct(StructData::try_new(
+            vec![StructField::new(
+                column_name!("operation_parameter_int").into_inner(),
+                DataType::INTEGER,
+                true,
+            )],
+            vec![Scalar::Null(DataType::INTEGER)],
+        )?)),
+        Expression::literal(format!("v{}", KERNEL_VERSION)),
+        column_expr!("engineCommitInfo"),
+    ];
+    let commit_info_expr = Expression::struct_from([Expression::struct_from(commit_info_exprs)]);
+
+    let engine_commit_info_schema = StructType::new(vec![StructField::new(
+        "engineCommitInfo",
+        MapType::new(DataType::STRING, DataType::STRING, true),
+        false,
+    )]);
+
+    // HACK (part 2/2): we need to modify the commit info schema to match the expression above (a
+    // struct with a single null int field).
+    let mut commit_info_empty_struct_schema = get_log_commit_info_schema().as_ref().clone();
+    let commit_info_field = commit_info_empty_struct_schema
+        .fields
+        .get_mut(COMMIT_INFO_NAME)
+        .ok_or_else(|| Error::missing_column(COMMIT_INFO_NAME))?;
+    let DataType::Struct(mut commit_info_data_type) = commit_info_field.data_type().clone() else {
+        return Err(Error::internal_error(
+            "commit_info_field should be a struct",
+        ));
+    };
+    commit_info_data_type
+        .fields
+        .get_mut("operationParameters")
+        .ok_or_else(|| Error::missing_column("operationParameters"))?
+        .data_type = DataType::Struct(Box::new(StructType::new(vec![StructField::new(
+        column_name!("hack_operation_parameter_int").into_inner(),
+        DataType::INTEGER,
+        true,
+    )])));
```

Comment:
That's really hard to read... can we pull out a `hack_data_type` and just assign it here?